librz/bin/pdb/tpi.c

```diff
@@ -951,1434 +563,1261 @@ static void free_snumeric(SNumeric *numeric) {
-	if (tmp_name) {
-		strcat(*name, tmp_name);
+	if (!rz_buf_read_le16(buf, &vftab->pad)) {
+		RZ_FREE(vftab);
+		return NULL;
 	}
-
-	//	if (need_to_free) {
-	//		free(tmp_name);
-	//		tmp_name = 0;
-	//	}
-}
-
-static void get_arglist_print_type(void *type, char **name) {
-	(void)type;
-	int name_len = 0;
-
-	name_len = strlen("arg_list");
-	*name = (char *)malloc(name_len + 1);
-	if (!(*name)) {
-		return;
-	}
-	// name[name_len] = '\0';
-	strcpy(*name, "arg_list");
-	//	STypeInfo *ti = (STypeInfo *) type;
-	//	SType *t = 0;
-	//	char *tmp_name = 0;
-	//	int name_len = 0;
-	//	int need_to_free = 1;
-	//	int base_type = 0;
-
-	//	base_type = ti->get_arg_type(ti, (void **)&t);
-	//	if (!t) {
-	//		need_to_free = 0;
-	//		print_base_type(base_type, &tmp_name);
-	//	} else {
-	//		ti = &t->type_data;
-	//		ti->get_print_type(ti, &tmp_name);
-	//	}
-
-	//	name_len = strlen("arglist ");
-	//	name_len += strlen(tmp_name);
-	//	*name = (char *) malloc(name_len + 1);
-	//	// name[name_len] = '\0';
-	//	strcpy(*name, "arglist ");
-	//	strcat(*name, tmp_name);
-
-	//	if (need_to_free)
-	//		free(tmp_name);
-}
-
-// TODO, nothing is really being parsed here
-static void get_mfunction_print_type(void *type, char **name) {
-	int name_len = 0;
-
-	name_len = strlen("mfunction ");
-	*name = (char *)malloc(name_len + 1);
-	if (!(*name)) {
-		return;
+	*read_len += sizeof(ut16);
+	if (!rz_buf_read_le32(buf, &vftab->index)) {
+		RZ_FREE(vftab);
+		return NULL;
 	}
-	// name[name_len] = '\0';
-	strcpy(*name, "mfunction ");
+	*read_len += sizeof(ut32);
+	return vftab;
 }
 
-static void get_union_print_type(void *type, char **name) {
-	STypeInfo *ti = (STypeInfo *)type;
-	//	ELeafType lt;
-	char *tmp_name = 0, *tmp1 = 0;
-	int name_len = 0;
-
-	//	lt = ti->leaf_type;
-	tmp_name = ti->get_name(ti);
-
-	tmp1 = "union ";
-	name_len = strlen(tmp1);
-	if (tmp_name) {
-		name_len += strlen(tmp_name);
+static Tpi_LF_Method *parse_type_method(RzBuffer *buf, ut16 len,
+	ut16 *read_len) {
+	rz_return_val_if_fail(buf, NULL);
+	Tpi_LF_Method *method = RZ_NEW0(Tpi_LF_Method);
+	if (!method) {
+		return NULL;
 	}
-	*name = (char *)malloc(name_len + 1);
-	if (!(*name)) {
-		return;
+	if (!rz_buf_read_le16(buf, &method->count)) {
+		RZ_FREE(method);
+		return NULL;
 	}
-	// name[name_len] = '\0';
-	strcpy(*name, tmp1);
-	if (tmp_name) {
-		strcat(*name, tmp_name);
+	*read_len += sizeof(ut16);
+	if (!rz_buf_read_le32(buf, &method->mlist)) {
+		RZ_FREE(method);
+		return NULL;
 	}
-
-	//	if (need_to_free) {
-	//		free(tmp_name);
-	//		tmp_name = 0;
-	//	}
+	*read_len += sizeof(ut32);
+	parse_type_string(buf, &method->name, len, read_len);
+	skip_padding(buf, len, read_len);
+	return method;
 }
 
-static void get_vtshape_print_type(void *type, char **name) {
-	int name_len = 0;
-
-	name_len = strlen("vtshape");
-	*name = (char *)malloc(name_len + 1);
-	if (!(*name)) {
-		return;
+static Tpi_LF_Member *parse_type_member(RzBuffer *buf, ut16 len,
+	ut16 *read_len) {
```

Comment:
I see this indentation in many places, please don't use so small line length limit, official limit is 100 chars: https://github.com/rizinorg/rizin/blob/dev/DEVELOPERS.md#code-style